llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -1848,7 +1850,95 @@ void lowerGlobalsToVector(Module &M) {
 
-PreservedAnalyses SYCLLowerESIMDPass::run(Module &M, ModuleAnalysisManager &) {
+bool SYCLLowerESIMDPass::prepareForAlwaysInliner(Module &M) {
+
+  auto markAlwaysInlined = [](Function &F) -> bool {
+    if (F.hasFnAttribute(llvm::Attribute::NoInline))
+      F.removeFnAttr(llvm::Attribute::NoInline);
+    if (F.hasFnAttribute(llvm::Attribute::InlineHint))
+      F.removeFnAttr(llvm::Attribute::InlineHint);
+    F.addFnAttr(llvm::Attribute::AlwaysInline);
+    return true;
+  };
+
+  bool NeedInline = false;
+  for (auto &F : M) {
+    // If some function already has 'alwaysinline' attribute, then request
+    // inliner pass.
+    if (F.hasFnAttribute(Attribute::AlwaysInline)) {
+      NeedInline = true;
+      continue;
+    }
+
+    // VC BE forbids 'alwaysinline' and "VCStackCall" on the same function.
+    // Such function may be used in other module, we cannot remove it
+    // after inlining.
+    if (F.hasFnAttribute(llvm::genx::VCFunctionMD::VCStackCall))
+      continue;
+
+    if (isESIMDKernel(F))
+      continue;
+
+    if (isSlmAllocatorConstructor(F) || isSlmAllocatorDestructor(F)) {
+      // slm_allocator constructor and destructor must be inlined
+      // to help SLM reservation analysis.
+      NeedInline |= markAlwaysInlined(F);
+      continue;
+    }
+
+    // TODO: The next code and comment was placed to ESIMDLoweringPass
+    // 2 years ago, when GPU VC BE did not support function calls and
+    // required everything to be inlined right into the kernel unless
+    // it had noinline or VCStackCall attrubute.
+    // This code migrated to here without changes, but... VC BE do support
+    //  the calls of spir_func these days, so this code may need re-visiting.
+    if (!F.hasFnAttribute(Attribute::NoInline))
+      NeedInline |= markAlwaysInlined(F);
+
+    if (!isSlmInit(F))
+      continue;
+
+    for (User *U : F.users()) {
+      auto *FCall = dyn_cast<CallBase>(U);
+      if (FCall && FCall->getCalledFunction() == &F) {
+        Function *GenF = FCall->getFunction();
+        // The original kernel (UserK) if often automatically separated into
+        // a spir_func (GenF) that is then called from spir_kernel (GenK).
+        // When that happens, the calls of slm_init<N>() originally placed
+        // in 'UserK' get moved to spir_func 'GenF', which creates wrong IR
+        // because slm_init() must be called only from a kernel.
+        // Fix it here: If 'GenF' has only 1 caller spir_kernel 'GenK',
+        // then inline 'GenF' to move slm_init call from spir_kernel 'GenK'.
+        SmallPtrSet<Function *, 1> GenFCallers;
+        for (User *GenFU : GenF->users()) {
+          auto *GenFCall = dyn_cast<CallBase>(GenFU);
+          if (GenFCall && GenFCall->getCalledFunction() == GenF) {
+            Function *GenK = GenFCall->getFunction();
+            GenFCallers.insert(GenK);
+          } else {
+            // Unexpected user of GenF. Do not require GenF inlining.
+            GenFCallers.clear();
```

Comment:
what about just returning false here?

Reply:
Do you mean 'return NeedInline', not 'return false' here, right?

Reply:
'return' cannot be called here because there is a loop for callers of slm_init().
Reaching this line only means that 'GenF' has some unexpected user/usage. It is Ok, then we just call 'break' and go to the next GenF (caller of slm_init)

Reply:
ah ok sorry i forgot about the outer loop